### PR TITLE
JSON errors parsing enhancement

### LINKED ISSUE #61: update to new Rust error format and/or move to json solely
We have 17 days until the old error format won't even be in stable anymore.

I'd be happy with moving to json entirely, it still has some quirks, but we could circumvent them for now by showing the json field that contains the full message in the popups.

### PR DESCRIPTION
**Pull request by alygin**
I've improved support for JSON error format and made it default.
1. All spans are now parsed and added to trace. No missing information anymore.
2. Label of the primary span is added to the main message. For instance, before: "mismatched types", after: "mismatched types (expected u32, found i32)".
3. Error code is added to the main message when possible. Before: "field `a` bound multiple times in the pattern", after: "field `a` bound multiple times in the pattern [E0027]".
4. For an errors within a macro, the macro definition is added to the trace. A special message type `Macro` is created for that.
5. A list of redundant messages is created to prevent displaying span labels that repeat the main message text in different wording. For instance, it's "struct `Foo` does not have a field named `bar`" instead of "struct `Foo` does not have a field named `bar` (struct `Foo` does not have field `bar`)". The list is safe to use because it checks exact message templates, so no useful information will be removed even if the error messages will be changed in the future. List is filled with three dozens of such messages, but can be easily expanded when redundancy is noticed.
6. If a trace item has the same location as the main message, this location is omitted in such an item to remove unnecessary noise.
7. A special case messages can be handled now before displaying. For instance, the compiler doesn't emit location for the error "main function not found". I add "src/main.rs" to such messages. It's not the ideal solution, but works better than what we have now.
8. Parsers (for JSON messages and for panics) are moved to their own files to simplify further maintenance.
9. JSON output format is default now.

I've checked it on several dozens of error messages and found no problems with missing useful information from the original JSON message. If you notice something, please, let me know.

I left the standard output parsing intact, just made a simple fix to prevent it from displaying broken messages when the new error format is used. So, it now works for the current error format but is effectively switched off for the new error format (I've provided a caveat in the "JSON format" setting description).

I'm intended to get back to the new error format parsing shortly. But it will be a separate PR.

There's an issue to discuss: what should we do with the existing settings? Should we forcefully switch all the users to the JSON format while upgrading? Since the standard output has been default so far, many use it and might lose linter messages after switching to the new error format if we do nothing. I'm not sure I'll have time to implement new parser quickly (after a closer look, I'm not even sure that it might be properly implemented for all errors).

Fixes #61

### CODE REVIEW

File: lib/errors.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  return level.charAt(0).toUpperCase() + level.slice(1);
+};
+
+// Set location for special cases when the compiler doesn't provide it
+function preprocessMessage(msg, buildWorkDir) {
+  if (msg.file) {
+    return true;
+  }
+  if (msg.message === 'main function not found') {
+    msg.file = buildWorkDir + '/src/main.rs';	// TODO: When running an example, set the path to the example file.
```
Comment (colin-kiegel) [inline lib/errors.js]: that path can be changed in `Cargo.toml` ... why not file it as a rustc bug, that the file is not reported in the json error represntation instead?

**Comment by colin-kiegel** [comment]
I think there is good reason to change the setting to json for everyone after the update

File: lib/errors.js (hunk 1 above)
Comment (alygin) [reply lib/errors.js]: I tried pointing it to `Cargo.toml` but then switched back because it's very unlikely that `Cargo.toml` is the file in which the developer will be fixing this problem. It's usually either `src/main.rs` or one of the examples. Making difference between `main.rs` and an example file requires passing a wider building context to this function. Since we don't do it now, I just left `main.rs`.

I'm not sure what solution is better. `Cargo.toml` is universal, but seems useless from the practical point of view.

Regarding filing a bug, I'm going to do that after studying it a bit closer to be able to come up with some practical suggestions. I [told Jonathan](https://internals.rust-lang.org/t/editor-compatibility-and-the-new-error-format/3435/32?u=alygin) about it but to no avail so far.

Comment (colin-kiegel) [reply lib/errors.js]: I suggest not to guess the filename in atom, if it is not specified. Instead you could display all errors with missing filenames as error pop-ups, as you did with the runtime errors.

 Ultimately this should be fixed in rustc instead.

Comment (alygin) [reply lib/errors.js]: I don't like guessing filenames either, but we only use it for a particular message for which it's really possible to guess the exact file. There's another exceptional error message: `error: aborting due to previous error` that is simply ignored now because it doesn't relate to the source code.

I would suggest the following:
1. Ignore the `aborting due to previous error` error.
2. Keep a special case for the `main function not found` error (provide the filename ourselves) until it's fixed in Rust ([here's the issue](https://github.com/rust-lang/rust/issues/36561)).
3. Display notifications for all other messages with no location in case we encounter something unknown at this time.

Comment (colin-kiegel) [reply lib/errors.js]: The thing is that the path can be customised, with the following `Cargo.toml` it is src/foo.rs and not src/main.rs.

`[[bin]]
name = "foo"
path = "src/foo.rs"`

You can completely overwrite this behaviour and even define multiple binaries. In these edge cases the guess will be wrong and point to a nonexistent or not the right file. Pointing to the wrong binary might be especially confusing if that one has a main function...

The question is, is it worth to have better diagnostics in 97℅ of the cases if that means getting 3℅ completely wrong? These numbers are just my personal estimate and may be far off.

Comment (alygin) [reply lib/errors.js]: You're right. Given such configuration possibilities, it's better not to divine. I'll remove this special case and add notifications.

**Comment by alygin** [comment]
Errors without locations are displayed as notifications now.
The "JSON error format" setting will be set to true for all users after the upgrade.

File: lib/json-parser.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+// of the given message
+function compareLocations(span, msg) {
+  return span.file_name === msg.file
+    && span.line_start === msg.line
+    && span.line_end === msg.line_end
+    && span.column_start === msg.col
+    && span.column_end === msg.col_end;
+}
+
+// Appends spans's lable to the main message. It only adds the lable if:
```
Comment (oli-obk) [inline lib/json-parser.js]: nit: label

File: lib/json-parser.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+// - the main message doesn't contain exactly the same phrase
+// - the main message doesn't contain the same information but uses different wording
+function appendSpanLabel(span, msg) {
+  if (!span.label || msg.message.indexOf(span.label) >= 0) {
+    return;
+  }
+  for (let idx = 0; idx < redundantLabels.length; idx++) {
```
Comment (oli-obk) [inline lib/json-parser.js:159]: `for (var l in redundantLabels)`

File: lib/panic-parser.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+'use babel';
+
+//
+// Panics and stack backtrades parser.
```
Comment (oli-obk) [inline lib/panic-parser.js]: nit: backtraces

**Review by oli-obk — CHANGES_REQUESTED** [review]
Just some nits. This is great work!

We should probably start bugging rustc devs to add a compiler option to generate json-producing panic code.

File: lib/json-parser.js (hunk 3 above)
Comment (alygin) [reply lib/json-parser.js:159]: In JS, `for (var l in redundantLabels)` iterates through indices, not through the array elements. At first glance, it looked counterintuitive to me, so I decided to leave the explicit iteration through indices.

`redundantLabels.forEach()` also doesn't work well here because it becomes too wordy when you need to return from the main function within a `forEach`-callback.

Comment (oli-obk) [reply lib/json-parser.js:159]: oh... that's very confusing. `redundandLabels.some()` is too unreadable, so lets stay with this.

File: lib/json-parser.js [hunk 5]
```diff
@@ -149,14 +149,14 @@ function compareLocations(span, msg) {
...
-// Appends spans's lable to the main message. It only adds the lable if:
+// Appends spans's label to the main message. It only adds the label if:
 // - the main message doesn't contain exactly the same phrase
 // - the main message doesn't contain the same information but uses different wording
 function appendSpanLabel(span, msg) {
   if (!span.label || msg.message.indexOf(span.label) >= 0) {
     return;
   }
-  for (let idx = 0; idx < redundantLabels.length; idx++) {
+  for (idx in redundantLabels) {
```
Comment (oli-obk) [inline lib/json-parser.js]: missing `var`

Comment (alygin) [reply lib/json-parser.js]: It's not only `var`, Travis requires an additional `if` within the loop because it thinks that I iterate through the simple object properties and must check if those properties are internal or not. With such check the loop becomes more wordy. It seems the original loop was more concise than any other option.

**Review by oli-obk — APPROVED** [review]
